README.md
# matrixeval-ruby

It's a plugin of [matrixeval](https://github.com/MatrixEval/matrixeval-ruby) for Ruby. Test your ruby code against multiple versions of dependencies like Ruby, Rails, Env ...

![](https://raw.githubusercontent.com/MatrixEval/assets/main/screenshots/summary.png)

## Features

- Parallel test your ruby code against multiple versions of dependencies combinations.
- Test your ruby code against a specific dependencies combination.
- Choose any docker image you like for each job.
- Easy to use CLI to speed up your development efficiency

## Requrements

You need to install docker and docker compose first

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'matrixeval-ruby'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install matrixeval-ruby

## Usage

Initialize

```bash
matrixeval init -t ruby
```

Customize `matrixeval.yml` file and run commands like:

```bash
matrixeval --all bundle install
matrixeval --all rspec
matrixeval --ruby 3.0 rspec a_spec.rb
matrixeval --ruby 3.1 --rails 7.0 rake test
matrixeval bash
```
Run `matrixeval --help` for more details

![](https://raw.githubusercontent.com/MatrixEval/assets/main/screenshots/help.png)

### Configuration Example

Here is the configuration file `matrixeval.yml` which will auto created by `matrixeval init`

```yaml
version: 0.4
target: ruby
project_name: REPLACE_ME
parallel_workers: number_of_processors
# commands:
#   - ps
#   - top
#   - an_additional_command
# mounts:
#   - /a/path/need/to/mount:/a/path/mount/to
matrix:
  ruby:
    variants:
      - key: 2.7
        container:
          image: ruby:2.7.1
      - key: 3.0
        default: true
        container:
          image: ruby:3.0.0
      - key: 3.1
        container:
          image: ruby:3.1.0
      # - key: jruby-9.3
      #   container:
      #     image: jruby:9.3
      #   env:
      #       PATH: "/opt/jruby/bin:/app/bin:/bundle/bin:$PATH"
      #   mounts:
      #     - /a/path/need/to/mount:/a/path/mount/to

  # rails:
  #   variants:
  #     - key: 6.1
  #       default: true
  #       env:
  #         RAILS_VERSION: "~> 6.1.0"
  #     - key: 7.0
  #       env:
  #         RAILS_VERSION: "~> 7.0.0"
  # another:
  #   variants:
  #     - key: key1
  #       default: true
  #       env:
  #         ENV_KEY: 1
  #     - key: key2
  #       env:
  #         ENV_KEY: 2

exclude:
  # - ruby: 3.0
  #   rails: 4.2
  # - ruby: jruby-9.3
  #   rails: 7.0

docker-compose-extend:
  # services:
  #   postgres:
  #     image: postgres:12.8
  #     volumes:
  #       - postgres12:/var/lib/postgresql/data
  #     environment:
  #       POSTGRES_HOST_AUTH_METHOD: trust

  #   redis:
  #     image: redis:6.2-alpine

  # volumes:
  #   postgres12:
```

### Gemfile configuration example

Here is an example from [ruby-trello](https://github.com/jeremytregunna/ruby-trello)

```ruby
if active_model_version = ENV['ACTIVE_MODEL_VERSION']
  gem 'activemodel', active_model_version
end
```

You can also check its corresponding [`matrixeval.yml`](https://github.com/jeremytregunna/ruby-trello/blob/master/matrixeval.yml)

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/matrixeval/matrixeval-ruby. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/[USERNAME]/matrixeval-ruby/blob/main/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the Matrixeval::Ruby project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/matrixeval/matrixeval-ruby/blob/main/CODE_OF_CONDUCT.md).


CHANGELOG.md
## [Unreleased]

## [0.4.0] - 2022-02-27

- Refactor this gem as a plugin of matrixeval

## [0.3.1] - 2022-02-21

- Add `.matrixeval/docker-compose` to `.gitignore`

## [0.3.0] - 2022-02-21

- Support add extra docker compose services and volumes
- Isolate each job with docker compose project name and network

## [0.2.2] - 2022-02-11

- Auto remove containers

## [0.2.1] - 2022-02-11

- Fix a assignment method issue

## [0.2.0] - 2022-02-10

- Change config format from 0.1 to 0.2
- CLI description
- Other errors fix

## [0.1.1] - 2022-02-09

- Fix a execution report issue

## [0.1.0] - 2022-02-09

- Initial release


test/matrixeval/ruby/target_test.rb
# frozen_string_literal: true

require "test_helper"

class Matrixeval::TargetTest < MatrixevalTest

  def setup
    @target = Matrixeval::Ruby::Target.new

    Matrixeval::Config::YAML.stubs(:yaml).returns({})
    ruby_vector = Matrixeval::Vector.new("ruby", {"key" => "ruby", "main" => true})
    sidekiq_vector = Matrixeval::Vector.new("sidekiq", {"key" => "sidekiq"})
    rails_vector = Matrixeval::Vector.new("rails", {"key" => "rails"})
    ruby_3_variant = Matrixeval::Variant.new({"key" => "3.0", "container" => { "image" => "ruby:3.0" }}, ruby_vector)
    sidekiq_5_variant = Matrixeval::Variant.new(
      {
        "key" => "5.0",
        "env" => {"SIDEKIQ_VERSION" => "5.0.0"}
      }, sidekiq_vector)

    rails_6_variant = Matrixeval::Variant.new(
      {
        "key" => "6.1",
        "env" => {"RAILS_VERSION" => "6.1.0"}
      }, rails_vector)

    @context = Matrixeval::Context.new(
      main_variant: ruby_3_variant,
      rest_variants: [sidekiq_5_variant, rails_6_variant]
    )
  end

  def test_version
    assert_equal '0.4.0', @target.version
  end

  def test_matrixeval_yml_template_path
    assert_match %r(lib/matrixeval/ruby/templates/matrixeval.yml), @target.matrixeval_yml_template_path.to_s
  end

  def test_vector_key
    assert_equal "ruby", @target.vector_key
  end

  def test_env
    expected_env = {
      "BUNDLE_PATH" => "/bundle",
      "GEM_HOME" => "/bundle",
      "BUNDLE_APP_CONFIG" => "/bundle",
      "BUNDLE_BIN" => "/bundle/bin",
      "PATH" => "/app/bin:/bundle/bin:$PATH"
    }
    assert_equal expected_env, @target.env(@context)
  end

  def test_mounts
    expected_mounts = [
      "bundle_ruby_3_0:/bundle",
      "../gemfile_locks/ruby_3_0_rails_6_1_sidekiq_5_0:/app/Gemfile.lock"
    ]
    assert_equal expected_mounts, @target.mounts(@context)
  end

  def test_volumes
    expected_volumes = {
      "bundle_ruby_3_0" => {
        "name" => "bundle_ruby_3_0"
      }
    }
    assert_equal expected_volumes, @target.volumes(@context)
  end

  def test_gitignore_paths
    assert_equal [".matrixeval/gemfile_locks"], @target.gitignore_paths
  end

  def test_support_commands
    expected_commands = [
      'ruby', 'rake', 'rails', 'rspec', 'bundle',
      'bin/rake', 'bin/rails', 'bin/rspec', 'bin/test'
    ]
    assert_equal expected_commands, @target.support_commands
  end

  def test_cli_example_lines
    expected_example_lines = [
      "",
      "Example:",
      "    matrixeval --all bundle install",
      "    matrixeval --ruby 3.0 rspec a_spec.rb",
      "    matrixeval --ruby 3.1 --active_model 7.0 rake test",
      "    matrixeval bash"
    ]
    assert_equal expected_example_lines, @target.cli_example_lines
  end

  def test_create_files
    Matrixeval.stubs(:working_dir).returns(dummy_gem_working_dir)
    FileUtils.rm_rf(dummy_gem_working_dir.join(".matrixeval/gemfile_locks"))
    Matrixeval::Context.stubs(:all).returns([@context])

    @target.create_files

    assert File.exist?(dummy_gem_working_dir.join(".matrixeval/gemfile_locks/ruby_3_0_rails_6_1_sidekiq_5_0"))
  end

end


lib/matrixeval/ruby/target.rb
module Matrixeval
  module Ruby
    class Target < Matrixeval::Target

      def version
        Matrixeval::Ruby::VERSION
      end

      def matrixeval_yml_template_path
        Matrixeval::Ruby.root.join("lib/matrixeval/ruby/templates/matrixeval.yml")
      end

      def vector_key
        "ruby"
      end

      def env(context)
        {
          "BUNDLE_PATH" => "/bundle",
          "GEM_HOME" => "/bundle",
          "BUNDLE_APP_CONFIG" => "/bundle",
          "BUNDLE_BIN" => "/bundle/bin",
          "PATH" => "/app/bin:/bundle/bin:$PATH"
        }
      end

      def mounts(context)
        bundle_volume = bundle_volume(context)

        [
          "#{bundle_volume}:/bundle",
          "../gemfile_locks/#{context.id}:/app/Gemfile.lock"
        ]
      end

      def volumes(context)
        bundle_volume = bundle_volume(context)

        {
          bundle_volume => {
            "name" => bundle_volume
          }
        }
      end

      def gitignore_paths
        [
          ".matrixeval/gemfile_locks"
        ]
      end

      def support_commands
        [
          'ruby', 'rake', 'rails', 'rspec', 'bundle',
          'bin/rake', 'bin/rails', 'bin/rspec', 'bin/test'
        ]
      end

      def cli_example_lines
        [
          "",
          "Example:",
          "    matrixeval --all bundle install",
          "    matrixeval --ruby 3.0 rspec a_spec.rb",
          "    matrixeval --ruby 3.1 --active_model 7.0 rake test",
          "    matrixeval bash"
        ]
      end

      def create_files
        gemfile_lock_folder = Matrixeval.working_dir.join(".matrixeval/gemfile_locks")
        FileUtils.mkdir_p gemfile_lock_folder

        Context.all.each do |context|
          FileUtils.touch gemfile_lock_folder.join(context.id)
        end
      end

      private

      def bundle_volume(context)
        docker_image = context.main_variant.container.image
        "bundle_#{docker_image.gsub(/[^A-Za-z0-9]/,'_')}"
      end
    end
  end
end
